cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/SingleThreadDatasetCache.java

```diff
@@ -280,25 +282,31 @@ public void dismissTransactionContext() {
 
   @Override
   public Iterable<TransactionAware> getTransactionAwares() {
-    if (txContext == null) {
-      return NO_TX_AWARES;
-    }
-    return Iterables.concat(extraTxAwares, activeTxAwares.values());
+    return (txContext == null) ? NO_TX_AWARES : Iterables.concat(activeTxAwares.values(), extraTxAwares);
   }
 
   @Override
   public void addExtraTransactionAware(TransactionAware txAware) {
-    extraTxAwares.add(txAware);
-    if (txContext != null) {
-      txContext.addTransactionAware(txAware);
+    // Use LIFO ordering. This method is only used for internal TransactionAware
+    // like Flow Queue publisher/consumer and TMS publisher/fetcher
+    // Using LIFO ordering is to allow TMS tx aware always be the last.
+    if (!extraTxAwares.contains(txAware)) {
+      extraTxAwares.addFirst(txAware);
+
+      // Starts the transaction on the tx aware if there is an active transaction
+      Transaction currentTx = txContext == null ? null : txContext.getCurrentTransaction();
+      if (currentTx != null) {
+        txAware.startTx(currentTx);
+      }
     }
   }
 
   @Override
   public void removeExtraTransactionAware(TransactionAware txAware) {
-    extraTxAwares.remove(txAware);
-    if (txContext != null) {
-      txContext.removeTransactionAware(txAware);
+    if (extraTxAwares.contains(txAware)) {
+      Preconditions.checkState(txContext == null || txContext.getCurrentTransaction() == null,
```

Comment:
It is needed because extra tx awares is not removing from the `TransactionContext`. Also we need to check if `txContext == null` first.